File: Java/2491. Divide Players Into Teams of Equal Skill.java
class Solution {
    public long dividePlayers(int[] skill) {
        int[] buckets = new int[1001];
        int min = 1001, max = 0;

        for (int s: skill) {
            min = Math.min(min, s);
            max = Math.max(max, s);
            buckets[s]++;
        }

        int sum = min + max;
        long ans = 0;

        for (int s: skill) {
            int partner = sum - s;

            if (buckets[partner] == 0) {
                return -1;
            }

            ans += s * partner;
            buckets[partner]--;
        }

        // Each element is counted twice
        return ans / 2;
    }
}

File: Python3/2491. Divide Players Into Teams of Equal Skill.py
class Solution:
    def dividePlayers(self, skill: List[int]) -> int:
        buckets = [0] * 1001
        smin, smax = 1001, 0

        for s in skill:
            buckets[s] += 1
            smin = min(smin, s)
            smax = max(smax, s)

        ssum = smin + smax
        ans = 0
        for s in skill:
            partner = ssum - s
            if buckets[partner] == 0:
                return -1

            ans += s * partner
            buckets[partner] -= 1

        return ans // 2